runtime/vm/ContinuationHelpers.cpp

```diff
@@ -923,14 +923,110 @@ preparePinnedVirtualThreadForUnmount(J9VMThread *currentThread, j9object_t syncO
+						lwEA = VM_ObjectMonitor::inlineGetLockAddress(currentThread, syncObj);
+						lock = J9_LOAD_LOCKWORD(currentThread, lwEA);
+						if (VM_ObjectMonitor::inlineFastInitAndEnterMonitor(currentThread, lwEA)) {
+							result = (UDATA)syncObj;
+							goto done;
+						}
+					}
+				}
+#endif /* J9VM_THR_LOCK_RESERVATION */
+				syncObjectMonitor = monitorTableAt(currentThread, syncObj);
+				monitor = syncObjectMonitor->monitor;
+
+				/* Try acquire the inflated monitor */
+				if (0 == omrthread_monitor_try_enter(monitor)) {
+					/* If the INFLATED bit is set, then it is already inflated and we own the object monitor. */
+					if (J9_ARE_ANY_BITS_SET(((J9ThreadMonitor *)monitor)->flags, J9THREAD_MONITOR_INFLATED)) {
+						currentThread->ownedMonitorCount += 1;
+						result = (UDATA)syncObj;
+						goto done;
+					}
+
+					/* Loop until either lock is inflated or FLC bit is set. */
+					while (!J9_LOCK_IS_INFLATED(lock)) {
+						/* The monitor isn't inflated yet, but we can update the lockword now. */
+						UDATA newLockword = 0;
+						UDATA count = J9_FLATLOCK_COUNT(lock);
+						if (0 == count) {
+							/* Lock is unlocked, so try to directly acquire it as a flatlock. */
+							newLockword = (UDATA)currentThread;
+						} else {
+							/* Change the lock to flat with FLC bit set so it will be inflated during exit. */
+							newLockword = ((UDATA)J9_FLATLOCK_OWNER(lock)
+										| ((count - 1) << OBJECT_HEADER_LOCK_V2_RECURSION_OFFSET)
+										| OBJECT_HEADER_LOCK_FLC);
+						}
+						j9objectmonitor_t const oldValue = lock;
+						lock = VM_ObjectMonitor::compareAndSwapLockword(currentThread, lwEA, lock, (j9objectmonitor_t)newLockword);
+						if (lock == oldValue) {
+							/* CAS succeeded, we can proceed with using the inflated monitor. */
+							VM_ObjectMonitor::incrementCancelCounter(J9OBJECT_CLAZZ(currentThread, syncObj));
+
+							if (J9_FLATLOCK_OWNER(lock) == currentThread) {
+								/* Lock is acquired. */
+								currentThread->ownedMonitorCount += 1;
+								result = (UDATA)syncObj;
+								goto done;
+							}
+							break;
+#if defined(J9VM_THR_LOCK_RESERVATION)
+						} else if (OBJECT_HEADER_LOCK_RESERVED == (lock & (OBJECT_HEADER_LOCK_RESERVED | OBJECT_HEADER_LOCK_INFLATED))) {
+							/* Lock is now reserved, exit the inflated monitor and restart to cancel lock reservation. */
+							omrthread_monitor_exit(monitor);
+							goto restart;
+#endif /* J9VM_THR_LOCK_RESERVATION */
+						}
+						/* CAS failed, another thread must have updated the lockword, retry the check. */
+					}
+				} else {
+					/* Inflated monitor owned by another thread, so the lockword update will be completed by them. */
+					lock = J9_LOAD_LOCKWORD(currentThread, lwEA);
+					/* We can safely continue if the lock is inflated or FLC bit is set. */
+					if (!J9_LOCK_IS_INFLATED(lock)
+					&& (OBJECT_HEADER_LOCK_FLC != (lock & (OBJECT_HEADER_LOCK_FLC | OBJECT_HEADER_LOCK_INFLATED)))
+					) {
```

Comment:
This can be simplified: `!J9_LOCK_IS_INFLATED(lock)` implies `OBJECT_HEADER_LOCK_INFLATED` is not set, in which case the second condition simplifies to `OBJECT_HEADER_LOCK_FLC != (lock & OBJECT_HEADER_LOCK_FLC)`. The two
conditions are true if and only if `0 == (lock & (OBJECT_HEADER_LOCK_FLC | OBJECT_HEADER_LOCK_INFLATED))` (which more closely matches the comment).
```c
					if (0 == (lock & (OBJECT_HEADER_LOCK_FLC | OBJECT_HEADER_LOCK_INFLATED))) {
```